File: Laborator/Java/AgentieTurism/src/main/java/ro/mpp2024/service/Service.java
package ro.mpp2024.service;

import ro.mpp2024.domain.Employee;
import ro.mpp2024.domain.Flight;
import ro.mpp2024.domain.Ticket;
import ro.mpp2024.repository.EmployeeRepository;
import ro.mpp2024.repository.FlightRepository;
import ro.mpp2024.repository.TicketRepository;
import ro.mpp2024.utils.FlightEvent;
import ro.mpp2024.utils.Observable;
import ro.mpp2024.utils.Observer;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

public class Service implements Observable<FlightEvent> {
    private final EmployeeRepository employeeRepo;
    private final FlightRepository flightRepo;
    private final TicketRepository ticketRepo;

    private final List<Observer<FlightEvent>> observers = new ArrayList<>();

    public Service(EmployeeRepository employeeRepo, FlightRepository flightRepo, TicketRepository ticketRepo) {
        this.employeeRepo = employeeRepo;
        this.flightRepo = flightRepo;
        this.ticketRepo = ticketRepo;
    }

    public Employee logIn(String email, String password) {
        return employeeRepo.logIn(email, password);
    }

    public List<Flight> getAllFlights() {
//        List<Flight> flightsList = new ArrayList<>();
//        flightRepo.findAll().forEach(flightsList::add);

        List<Flight> flightsList = flightRepo.findFlights();
        return flightsList;
    }

    public List<Flight> findFlights(String destination, LocalDateTime departureDateTime) {
        return flightRepo.findByDestinationAndDateTime(destination, departureDateTime);
    }

    public void buyTickets(Flight flight, String clientName, String turistsName, String clientAddress, int seatsNumber) {
        int availableSeats = flight.getAvailableSeats();

        if (seatsNumber > availableSeats) {
            throw new RuntimeException("Nu sunt suficiente locuri disponibile!");
        }

        Ticket ticket = new Ticket(flight, clientName, turistsName, clientAddress, seatsNumber);
        ticketRepo.add(ticket);

        int newSeats = availableSeats - seatsNumber;
        flightRepo.updateAvailableSeats(flight.getId(), newSeats);

        notifyObservers(new FlightEvent(FlightEvent.EventType.FLIGHTS_UPDATED));
    }


    @Override
    public void addObserver(Observer<FlightEvent> observer) {
        observers.add(observer);
    }

    @Override
    public void removeObserver(Observer<FlightEvent> observer) {
        observers.remove(observer);
    }

    @Override
    public void notifyObservers(FlightEvent event) {
        for (Observer<FlightEvent> observer : observers) {
            observer.update(event);
        }
        printRed("Notific observerii...");
    }

    public List<Observer<FlightEvent>> getObservers() {
        return observers;
    }

    public static void printRed(String message) {
        System.out.println("\u001B[31m" + message + "\u001B[0m");
    }
}

File: Laborator/CSharp/AgentieTurism/Service/Service.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using AgentieTurism.Models;
using AgentieTurism.Repository;
using AgentieTurism.Utils;
using log4net;

namespace AgentieTurism.Services
{
    public class Service : Observable<FlightEvent>
    {
        private static readonly ILog log = LogManager.GetLogger(typeof(Service));
        private readonly IEmployeeRepository employeeRepo;
        private readonly IFlightRepository flightRepo;
        private readonly ITicketRepository ticketRepo;
        private readonly List<Observer<FlightEvent>> observers = new List<Observer<FlightEvent>>();

        public Service(IEmployeeRepository employeeRepo, IFlightRepository flightRepo, ITicketRepository ticketRepo)
        {
            log.Info("Initializing Service");
            this.employeeRepo = employeeRepo;
            this.flightRepo = flightRepo;
            this.ticketRepo = ticketRepo;
        }

        public Employee LogIn(string email, string password)
        {
            log.Info($"Logging in user with email: {email}");
            return employeeRepo.LogIn(email, password);
        }

        public List<Flight> GetAllFlights()
        {
            log.Info("Retrieving all flights");
            return flightRepo.FindFlights();
        }

        public List<Flight> FindFlights(string destination, DateTime departureDateTime)
        {
            log.Info($"Searching flights for destination: {destination} on {departureDateTime}");
            return flightRepo.FindByDestinationAndDateTime(destination, departureDateTime);
        }

        public void BuyTickets(Flight flight, string clientName, string turistsName, string clientAddress, int seatsNumber)
        {
            log.Info($"Attempting to buy {seatsNumber} tickets for {clientName} on flight {flight.Id}");

            int availableSeats = flight.AvailableSeats;
            if (seatsNumber > availableSeats)
            {
                throw new Exception("Not enough seats available!");
            }

            Ticket ticket = new Ticket(flight, clientName, turistsName, clientAddress, seatsNumber);
            ticketRepo.Add(ticket);

            int newSeats = availableSeats - seatsNumber;
            flightRepo.UpdateAvailableSeats(flight.Id, newSeats);

            NotifyObservers(new FlightEvent(FlightEvent.EventType.FLIGHTS_UPDATED));
        }

        public void AddObserver(Observer<FlightEvent> observer)
        {
            observers.Add(observer);
        }

        public void RemoveObserver(Observer<FlightEvent> observer)
        {
            observers.Remove(observer);
        }

        public void NotifyObservers(FlightEvent eventObj)
        {
            foreach (var observer in observers)
            {
                observer.Update(eventObj);
            }
            Console.WriteLine("Notifying observers...");
        }
    }
}
